### PR TITLE
prevent TS from creating sessions for repo status checks

### PR DESCRIPTION
**Pull request by martenson**
this has been done:
- enable Routes middleware
- redirect in the middleware before the request reaches the app
- properly forward the contents of the query string
- move the controller method to the API namespace and properly decorate
- provide the ability to not hexlify the response (so we can dump it in future Galaxy release)

I hope that this work will allow for **easier transition from controllers to APIs** and also **smoother deprecation of API endpoints** in the future.

Thanks to @natefoo for help.

might be of interest to @carlfeberhard @guerler @dannon @jmchilton

### CODE REVIEW

**Comment by martenson** [comment]
`funny` stat: we have ~250 000 000 sessions in MTS's DB

**Comment by dannon** [comment]
This works well for me locally, are those ts test failures legit?

**Comment by martenson** [comment]
@dannon dunno, it might be conflicting middleware, we have had something similar before :/

**Comment by natefoo** [comment]
Correction: We _had_ ~250,000,000 sessions:

``` sql
galaxy_toolshed=# delete from galaxy_session where remote_addr in (select remote_addr from galaxy_session group by remote_addr having count(remote_addr) > 10000);
DELETE 198819643
```

Thanks @martenson!

**Comment by dannon** [comment]
Hope postgres remembers how to vacuum.

**Comment by natefoo** [comment]
For the TS test failures, try loading the routes middleware after the hg middleware.

**Comment by natefoo** [comment]
@dannon:

```
galaxy_toolshed=# vacuum verbose galaxy_session;
INFO:  vacuuming "public.galaxy_session"
```

Waiting... =)

**Comment by martenson** [comment]
well, this did not help @natefoo

**Comment by carlfeberhard** [comment]
Might be good/more RESTful to make the endpoint a noun like: `/api/repositories/updates`. Otherwise, looks great to me +1

**Comment by martenson** [comment]
done, thanks @carlfeberhard

**Comment by dannon** [comment]
@martenson I'm going to try a few things locally to (re)wrap the middleware around the application so that it works, unless you've already fixed it.

**Comment by martenson** [comment]
@dannon help appreciated, I will not work on this until tomorrow

File: lib/galaxy/webapps/tool_shed/buildapp.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import atexit
 import logging
 import os
+import urlparse
```
Comment (nsoranzo) [inline lib/galaxy/webapps/tool_shed/buildapp.py]: `from six.moves.urllib.parse import parse_qs`

File: lib/galaxy/webapps/tool_shed/buildapp.py [hunk 2]
```diff
@@ -284,3 +299,19 @@ def wrap_in_middleware( app, global_conf, **local_conf ):
...
+
+
+def _map_redirects( app ):
+    """
+    Add redirect to the Routes mapper and forward the received query string.
+    Subsequently when the redirect is triggered in Routes middleware the request
+    will not even reach the webapp.
+    """
+    def forward_qs(environ, result):
+        qs_dict = urlparse.parse_qs(environ['QUERY_STRING'])
```
Comment (nsoranzo) [inline lib/galaxy/webapps/tool_shed/buildapp.py]: s/urlparse.parse_qs/parse_qs/

**Comment by martenson** [comment]
routes.Middleware is somehow conflicting with the hgwebapp we use for cloning and for the hg browser. The `GET /repos/qqqqqq/asdsadsadas` which is supposed to show mercurial browser of the repo asdsadsadas of owner qqqqqq will end up doing this (404):
note: the hg.py controller is never actually entered here

```
routes.middleware DEBUG 2016-08-15 15:04:23,080 Matched GET /repos/qqqqqq/asdsadsadas
routes.middleware DEBUG 2016-08-15 15:04:23,080 Route path: 'repos/*path_info', defaults: {'action': u'handle_request', 'controller': u'hg', 'path_info': u'/'}
routes.middleware DEBUG 2016-08-15 15:04:23,081 Match dict: {'action': u'handle_request', 'controller': u'hg', 'path_info': 'qqqqqq/asdsadsadas'}
128.118.201.178 - - [15/Aug/2016:15:04:23 -0400] "GET /repos/qqqqqq/asdsadsadas HTTP/1.1" 404 - "http://martenson.bx.psu.edu:9009/repository/manage_repository?sort=name&operation=view_or_manage_repository&id=75cfda0e4b8063b8" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36"
```

instead of this (resources form the hgwebapp):

```
128.118.201.178 - - [15/Aug/2016:16:09:22 -0400] "GET /repos/qqqqqq/asdsadsadas HTTP/1.1" 200 - "http://martenson.bx.psu.edu:9009/repository/manage_repository?sort=name&operation=view_or_manage_repository&id=75cfda0e4b8063b8" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36"
128.118.201.178 - - [15/Aug/2016:16:09:23 -0400] "GET /repos/qqqqqq/asdsadsadas/static/style-paper.css HTTP/1.1" 200 6898 "http://martenson.bx.psu.edu:9009/repos/qqqqqq/asdsadsadas" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36"
128.118.201.178 - - [15/Aug/2016:16:09:23 -0400] "GET /repos/qqqqqq/asdsadsadas/static/hglogo.png HTTP/1.1" 200 4123 "http://martenson.bx.psu.edu:9009/repos/qqqqqq/asdsadsadas" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36"
128.118.201.178 - - [15/Aug/2016:16:09:23 -0400] "GET /repos/qqqqqq/asdsadsadas/static/feed-icon-14x14.png HTTP/1.1" 200 689 "http://martenson.bx.psu.edu:9009/repos/qqqqqq/asdsadsadas" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36"
128.118.201.178 - - [15/Aug/2016:16:09:23 -0400] "GET /repos/qqqqqq/asdsadsadas/static/mercurial.js HTTP/1.1" 200 12264 "http://martenson.bx.psu.edu:9009/repos/qqqqqq/asdsadsadas" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36"
```

I do not know how to overcome this yet.

**Comment by natefoo** [comment]
How is the routes middleware aware of the route match inside the app?

File: lib/galaxy/webapps/tool_shed/buildapp.py [hunk 3]
```diff
@@ -199,11 +210,15 @@ def wrap_in_middleware( app, global_conf, **local_conf ):
...
     conf.update( local_conf )
     debug = asbool( conf.get( 'debug', False ) )
+    routes_mapper = app.mapper
     # First put into place httpexceptions, which must be most closely
     # wrapped around the application (it can interact poorly with
     # other middleware):
     app = httpexceptions.make_middleware( app, conf )
     log.debug( "Enabling 'httpexceptions' middleware" )
+    # Load the Routes middleware which we use for redirecting
+    app = RoutesMiddleware( app, routes_mapper )
```
Comment (martenson) [inline lib/galaxy/webapps/tool_shed/buildapp.py]: @natefoo it is being passed the mapper here

**Comment by martenson** [comment]
I might have found a way around this creating a special mapper for the redirects. Will explore more tomorrow.

**Comment by martenson** [comment]
closing for now, Jenkins testing on local branch

**Comment by martenson** [comment]
@natefoo passing!

**Comment by dannon** [comment]
Working for me locally too, nice!  I'll merge once Travis is done.

**Comment by martenson** [comment]
@natefoo did you identify other endpoints that we should treat in the same manner?

**Comment by nsoranzo** [comment]
@martenson Can you fix the 2 small flake8 errors in https://travis-ci.org/galaxyproject/galaxy/jobs/152742479 ?

**Comment by natefoo** [comment]
@martenson Here are the top hits:

```
  12252 /repository/get_tool_versions
  13757 /repository/next_installable_changeset_revision
  21291 /api/repositories/get_repository_revision_install_info
  34620 /repository/previous_changeset_revisions
  50046 /repository/get_required_repo_info_dict
  67879 /repository/get_repository_dependencies
 136300 /api/repositories/get_ordered_installable_revisions
 180963 /repository/get_tool_dependencies
1908410 /repository/status_for_installed_repository
```

**Comment by martenson** [comment]
@natefoo seems we solved ~80% with that one stone, do we want to treat `/repository/get_tool_dependencies` the same or are we good for now?

**Pull request merged**